Follow-up of #71136: Switch system priority class usage to versioned (v1) api

yue9944882 opened the pull request:
/kind cleanup
/cc @liggitt

Fixes https://github.com/kubernetes/kubernetes/issues/84246

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

yue9944882 committed:
switch system priority class to versioned (v1) api

move all the helpers to scheduling v1 helpers

less explicit conversion

yue9944882 commented:
/retest

liggitt commented on pkg/apis/scheduling/v1/helpers.go:
```diff
@@ -0,0 +1,66 @@
+/*
```
can we remove the pkg/apis/scheduling/helpers.go file?

yue9944882 replied:
we're still using the helper funcs in the validation here: 

https://github.com/kubernetes/kubernetes/blob/ef830e2bcf7a23a0ae74e8d8b9e8811a83142f38/pkg/apis/scheduling/validation/validation.go#L37-L39

removing the file will make us do a explicit conversion to v1 api here which looks a bit odd to me. wdyt?

liggitt replied:
no conversion needed, that method is checking name (string), value (int), and default (bool) matches

liggitt commented:
one comment on removing the internal-typed helpers.go file, lgtm otherwise

cc @bsalamat @kubernetes/sig-scheduling-pr-reviews

ravisantoshgudimetla commented:
Thanks for the PR @yue9944882. 

I am ok with helpers too but I will wait for Bobby to give his review.

/cc @bsalamat 

/approve

yue9944882 commented:
thanks @ravisantoshgudimetla , ping @bsalamat for review

bsalamat commented on pkg/apis/scheduling/v1/helpers.go:
```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
2019

bsalamat commented on staging/src/k8s.io/api/scheduling/v1alpha1/types.go:
```diff
@@ -20,6 +20,27 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+const (
```
Do we need these consts in v1alpha1 and v1beta1?

yue9944882 replied:
i removed them all from the external apis. now there's no copy of these consts.

liggitt commented:
can we remove the SystemPriorityClasses() function from pkg/apis/scheduling/helpers.go and switch all usage to the v1 version?

yue9944882 commented:
@liggitt now i switched the helpers to v1 completely and removed the internal ones. but there's still usage of internal helpers left in the priority class validation, i switched them too, not sure if it looks good.

https://github.com/kubernetes/kubernetes/blob/1409ff38e5828f55bf0e57fa32f93f6ef5b63a25/pkg/apis/scheduling/validation/validation.go#L37

bsalamat commented on pkg/apis/scheduling/v1/helpers.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2018 The Kubernetes Authors.
+Copyright 2019 The Kubernetes Authors.
```
Do we need to change the copyright year?

yue9944882 replied:
mmm am not sure, actually i was copy-pasting the versioned apis to a new pkg but the github insists it's a renaming 🤷🏻‍♂️

bsalamat commented on pkg/apis/scheduling/v1/helpers.go:
```diff
@@ -14,43 +14,62 @@ See the License for the specific language governing permissions and
+
+// IsKnownSystemInternalPriorityClass checks that "pc" is equal to one of the system PriorityClasses.
+// It ignores "description", labels, annotations, etc. of the PriorityClass.
+func IsKnownSystemInternalPriorityClass(pc *scheduling.PriorityClass) (bool, error) {
```
Why do we need to distinguish "internal" from other system priority classes?

yue9944882 replied:
it's the validation of priority-class using the internal helpers so we can't completely strip this.

bsalamat replied:
I am not sure if I understand what problem you are trying to solve here. It looks like `IsKnownSystemPriorityClass` is no longer used. Couldn't have we made the necessary changes to  `IsKnownSystemPriorityClass` and not added `IsKnownSystemInternalPriorityClass`?

yue9944882 replied:
the both is being used. i'm trying to avoid doing explicit conversion between internal types and versioned types. 

`IsKnownSystemPriorityClass` is in the post-start-hook, and `IsKnownSystemInternalPriorityClass` is in the validation [https://github.com/kubernetes/kubernetes/blob/eaf89cfbb483f99411141f70c294e8a014e49cec/pkg/apis/scheduling/validation/validation.go#L37](https://github.com/kubernetes/kubernetes/blob/eaf89cfbb483f99411141f70c294e8a014e49cec/pkg/apis/scheduling/validation/validation.go#L37)

liggitt replied:
Alternately, make IsKnownSystemPriorityClass take the name, value, globalDefault and preemptionPolicy fields as parameters, rather than the entire object. Then both callers can collapse onto a single method

liggitt commented:
This is needed to allow the API server to start with the beta API version turned off

liggitt commented:
/lgtm
/approve
/retest

ahg-g commented:
This may have broke scheduler_perf:

ahg@ahg1:~/go/src/k8s.io/kubernetes$ make test-integration WHAT=./test/integration/scheduler_perf KUBE_TEST_VMODULE="''" KUBE_TEST_ARGS="-alsologtostderr=false -logtostderr=false -run=^$$ -bench=BenchmarkScheduling/5000Nodes/1000Pods -cpuprofile cpu.out" &> /tmp/test

```
make[1]: *** No rule to make target '/usr/local/google/home/ahg/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/apis/scheduling/helpers.go', needed by 'pkg/apis/scheduling/zz_generated.deepcopy.go'.  Stop.
make: *** [Makefile:559: generated_files] Error 2

```

liggitt commented:
I cannot reproduce on master with a clean checkout:

```
$ git show
commit 5238e1c80fea02891e1804b346d24faa7c13da07 (HEAD -> master, upstream/master, upstream/HEAD, origin/master)
Merge: 31b7e3d9053 09cf42d67c0

$ make clean

$ make test-integration WHAT=./test/integration/scheduler_perf KUBE_TEST_VMODULE="''" KUBE_TEST_ARGS="-alsologtostderr=false -logtostderr=false -run=^$$ -bench=None -cpuprofile cpu.out"
```

> ```
> +++ [1024 08:52:18] Building go targets for darwin/amd64:
>     ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
> +++ [1024 08:52:30] Building go targets for darwin/amd64:
>     ./vendor/k8s.io/code-generator/cmd/defaulter-gen
> +++ [1024 08:52:38] Building go targets for darwin/amd64:
>     ./vendor/k8s.io/code-generator/cmd/conversion-gen
> +++ [1024 08:52:52] Building go targets for darwin/amd64:
>     ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
> +++ [1024 08:53:06] Building go targets for darwin/amd64:
>     ./vendor/github.com/go-bindata/go-bindata/go-bindata
> +++ [1024 08:53:08] Checking etcd is on PATH
> /Users/liggitt/go/src/k8s.io/kubernetes/third_party/etcd/etcd
> +++ [1024 08:53:08] Starting etcd instance
> etcd --advertise-client-urls http://127.0.0.1:2379 --data-dir /var/folders/h8/3dzrjpfj76d93vbhc5f2r8tw00kjgb/T/tmp.22gcXzY5 --listen-client-urls http://127.0.0.1:2379 --debug > "/dev/null" 2>/dev/null
> Waiting for etcd to come up.
> +++ [1024 08:53:09] On try 1, etcd: : http://127.0.0.1:2379
> {"action":"set","node":{"key":"/_test","value":"","modifiedIndex":4,"createdIndex":4}}
> +++ [1024 08:53:09] Running integration test cases
> ```

ahg-g commented:
Thanks Jordan, doing ```make clean``` solved the issue for me.

josiahbjorgaard commented:
Bug triage for 1.17 here. Code freeze is Nov. 18th. Will this PR be merged by then? If not, can we move it out of the milestone?
